File: 39.py
class Solution(object):
    def combinationSum(self, candidates, target):
        
        def dfs(remain, combo, index):
            if remain < 0:
                return
            if remain == 0:
                result.append(combo)
                return
            for i in range(index, len(candy)):                
                print(i, combo)
                dfs(remain - candy[i], combo + [candy[i]], i)
                
        candy = sorted(candidates)
        result = []
        dfs(target, [], 0)
        return result

print(Solution().combinationSum([10,1,2,7,6,1,5], 8))

File: 39.go
package main
import (
    "fmt"
    "sort"
    "reflect"
)
func combinationSum(candidates []int, target int) [][]int {
    results := [][]int{}
    sequence := []int{}
    sort.Ints(candidates)
    sum(target, sequence,0, &results, &candidates)
//    for _,slice := range results{
//        fmt.Println(slice)
//    }
    return results
}

var depth = 0
func sum(target int, sequence []int, index int, results *[][]int, candidates *[]int){
    depth += 1
    if target < 0{
        return
    } else if target == 0 {
        addResult(results, sequence)
        return
    }else{
        for i:=index; i<len(*candidates); i++{
            num := (*candidates)[i]
            //sequence = append(sequence, num)
            sum(target-num, append(sequence,num), i, results, candidates)
            //sequence = sequence[:len(sequence)-1]
        }
    }
}

func addResult(results *[][]int , sequence []int){
    sort.Ints(sequence)
    for _,slice := range *results{
        if reflect.DeepEqual(slice, sequence) {
            return
        }
    }
    result := make([]int, len(sequence))
    copy(result, sequence)
    *results = append(*results, result)
}

func main(){
    candidates := []int{2,3,7}
    combinationSum(candidates, 18)
}
